include/ginkgo/core/matrix/row_gatherer.hpp

```diff
@@ -0,0 +1,183 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MATRIX_ROW_GATHERER_HPP_
+#define GKO_PUBLIC_CORE_MATRIX_ROW_GATHERER_HPP_
+
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <vector>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * RowGatherer is a matrix "format" which stores the gather indices arrays which
+ * can be used to gather rows to another matrix.
+ *
+ * @tparam IndexType  precision of rowgatherer array indices.
+ *
+ * @note This format is used mainly to allow for an abstraction of the
+ * rowgatherer and provides the user with an apply method which
+ * calls the respective LinOp's rowgatherer operation if the respective LinOp
+ * implements the RowGatherable interface. As such it only stores an array of
+ * the rowgatherer indices.
+ *
+ * @ingroup rowgatherer
+ * @ingroup matrix
+ * @ingroup LinOp
+ */
+template <typename IndexType = int32>
+class RowGatherer : public EnableLinOp<RowGatherer<IndexType>>,
+                    public EnableCreateMethod<RowGatherer<IndexType>> {
+    friend class EnableCreateMethod<RowGatherer>;
+    friend class EnablePolymorphicObject<RowGatherer, LinOp>;
+
+public:
+    using index_type = IndexType;
+
+    /**
+     * Returns a pointer to the array of row_gatherer_indices.
+     *
+     * @return the pointer to the row_gatherer_indices array.
+     */
+    index_type* get_row_gatherer_indices() noexcept
+    {
+        return row_gatherer_indices_.get_data();
+    }
+
+    /**
+     * @copydoc get_row_gatherer_indices()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type* get_const_row_gatherer_indices() const noexcept
+    {
+        return row_gatherer_indices_.get_const_data();
+    }
+
+    /**
+     * Returns the number of elements explicitly stored in the
+     * row_gatherer_indices array.
+     *
+     * @return the number of elements explicitly stored in the
+     * row_gatherer_indices array.
+     */
+    size_type get_row_gatherer_indices_size() const noexcept
+    {
+        return row_gatherer_indices_.get_num_elems();
+    }
```

Comment:
That's equal to get_size()[0], right? Why do we need a separate function for that?

Reply:
that's true. I generate this from permutation, so just rename it to fit row_gather.
I will remove it